cmd/harmony/main.go

```diff
@@ -699,6 +708,37 @@ func setupPrometheusService(node *node.Node, hc harmonyConfig, sid uint32) {
 	node.RegisterService(service.Prometheus, p)
 }
 
+func setupSyncService(node *node.Node, host p2p.Host, hc harmonyConfig) {
+	blockchains := []*core.BlockChain{
+		node.Blockchain(),
+		node.Beaconchain(),
+	} // The duplicate blockchain will be skipped in config parsing
+	dConfig := downloader.Config{
+		Network:      nodeconfig.NetworkType(hc.Network.NetworkType),
+		Concurrency:  hc.Sync.Concurrency,
+		MinStreams:   hc.Sync.MinPeers,
+		InitStreams:  hc.Sync.InitStreams,
+		SmSoftLowCap: hc.Sync.DiscSoftLowCap,
+		SmHardLowCap: hc.Sync.DiscHardLowCap,
+		SmHiCap:      hc.Sync.DiscHighCap,
+		SmDiscBatch:  hc.Sync.DiscBatch,
+	}
+	// If we are running side chain, we will need to do some extra works for beacon
+	// sync
+	if node.NodeConfig.ShardID != 0 {
+		dConfig.BHConfig = &downloader.BeaconHelperConfig{
+			BlockC:     node.BeaconBlockChannel,
+			InsertHook: node.BeaconSyncHook,
```

Comment:
where is this Hook used? why we need this hook?

Reply:
This hook is to broadcast the beacon crosslink if the node is a side-chain leader. It is called in `downloader.BeaconHelper`.

This hook is actually making cross-chain transaction possible. Without it, the cross-shard transaction will be blocked for lacking knowledge of crosslink data. 

And this is the refactor of the old DNS sync logic:

```
// DoBeaconSyncing update received beaconchain blocks and downloads missing beacon chain blocks
func (node *Node) DoBeaconSyncing() {
	if node.NodeConfig.IsOffline {
		return
	}

	go func(node *Node) {
		// TODO ek – infinite loop; add shutdown/cleanup logic
		for beaconBlock := range node.BeaconBlockChannel {
			if node.beaconSync != nil {
				err := node.beaconSync.UpdateBlockAndStatus(
					beaconBlock, node.Beaconchain(), node.BeaconWorker, true,
				)
				if err != nil {
					node.beaconSync.AddLastMileBlock(beaconBlock)
				} else if node.Consensus.IsLeader() {
					// Only leader broadcast crosslink to avoid spamming p2p
					node.BroadcastCrossLink()  <---------------------------------------NOTE THIS LINE
				}
			}
		}
	}(node)
```

Reply:
cross-shard transaction doesn't depends on cross-links. Cross-shard txns only depends on the correct committee information to verify the signatures.

Reply:
Actually, the local rpc tests fails at CX transactions before I added this piece of code. Let me dig deeper into the mechanism then.

Reply:
Probably the communication between shard is broken. Though crosslink is not needed. The cross-shard beacon block broadcasting is still needed for other shard to receive the new committee info in time.

Reply:
Cross-shard beacon block broadcasting is fine as far as I see. You can start running localnet and check the side chain's epoch changes swiftly as same as beacon epoch, which means beacon committee is synced OK.

Reply:
keep in mind that local cx txn testing is flaky. So it may not be a problem in the code.